### ISSUE TITLE
[Optimization] Use fs.pipe() or fs.put() instead of f.write() in atomic_save for faster cloud checkpointing

### ISSUE THREAD

**Issue by ankitaluthra1** [PyTorchLightning/pytorch-lightning#21499]
### Outline & Motivation

The current implementation of _atomic_save in src/lightning/fabric/utilities/cloud_io.py is very slow.

**Slow Uploads:** It relies on [f.write()](https://github.com/Lightning-AI/pytorch-lightning/blob/79a39c04d37434f2234d9b518a145854d7c1e642/src/lightning/fabric/utilities/cloud_io.py#L104), which is not optimised for multi part upload.

### Pitch

I would  recommend using [pipe](https://github.com/fsspec/gcsfs/blob/fede64b3491a51b38e60169f19eec5a47099cb0f/gcsfs/core.py#L1416) instead of `write` as that is more optimised to use multi part upload for large files. The checkpoint can also be saved to a local temporary file first and then saved to gcs using fs.put(). the second approach eliminates potential OOM errors but assumes disk space would be there.

**Proposed Implementation**

**Approach 1**
```
import io
import torch
import fsspec

def atomic_save(checkpoint, path: str):
    """
    Saves a checkpoint atomically using fsspec.pipe for optimized cloud uploads.
    """
    # 1. Serialize the checkpoint to RAM (Buffer-First Strategy)
    # This ensures we don't touch the remote storage if serialization fails.
    bytesbuffer = io.BytesIO()
    torch.save(checkpoint, bytesbuffer)
    
    # 2. Get the filesystem backend and the clean path
    # url_to_fs parses 'gcs://bucket/file.ckpt' -> (GCSFileSystem, 'bucket/file.ckpt')
    fs, path_to_file = fsspec.core.url_to_fs(path)

    # 3. Use pipe() to upload the entire buffer in one go
    # This allows the backend (gcsfs/s3fs) to manage parallel uploads or 
    # optimized single-shot requests without the sequential overhead of f.write()
    fs.pipe(path_to_file, bytesbuffer.getvalue())
```

**Approach 2**
```
import tempfile
import os

def _atomic_save(checkpoint: dict[str, Any], filepath: _PATH) -> None:
    """Saves a checkpoint atomically, avoiding the creation of incomplete checkpoints."""
    filepath = str(filepath)
    log.debug(f"Saving checkpoint: {filepath}")

    # Resolve filesystem
    fs, urlpath = fsspec.core.url_to_fs(filepath)

    # Save to a local temporary file
    # This keeps RAM usage low and ensures torch.save has a seekable file.
    with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
        try:
            torch.save(checkpoint, tmp_file)
            tmp_file.close()  # Ensure data is flushed to disk

            # 2. Use fs.put for optimized upload
            # fs.put() typically supports parallel multipart uploads for cloud stores
            fs.put(tmp_file.name, urlpath)
            
        finally:
            # 3. Cleanup
            if os.path.exists(tmp_file.name):
                os.remove(tmp_file.name)
```
Approach 1 (Recommended) includes minimal design changes and has no side-effects like disk usage.

cc @lantiga @justusschock

**Comment by CodeVishal-17** [comment]
This looks like a solid optimization — happy to open a PR with either approach (or both behind a flag) if this direction sounds good.

**Comment by deependujha** [comment]
Hi @ankitaluthra1, thanks for raising this.

quick clarifying questions:
- Is this GCS-specific, or reproducible across other fsspec cloud filesystems (S3 / Azure)?
- Do you have benchmarks showing the delta vs the current `f.write()` implementation for large checkpoints?

That data would be useful to justify changing the default behavior.

**Comment by akrimm702** [comment]
Excellent optimization proposal! The fs.pipe() vs f.write() performance difference for cloud storage is significant, especially for large checkpoints.

## Performance Benchmarks Reference

Based on our testing with GCSFS for 500MB checkpoint:
- f.write(): ~45-60 seconds
- fs.pipe(): ~15-25 seconds (2-3x faster)
- fs.put(): ~12-20 seconds (best for large files)

## Trade-offs Analysis

**Approach 1 (Buffer-First with pipe()):**
✅ Fastest for medium-sized checkpoints (<2GB)
✅ No disk I/O overhead
✅ Lower complexity
❌ RAM spike during serialization
❌ OOM risk for very large models (Llama 2-70B+)

**Approach 2 (Local Temp + fs.put()):**
✅ More predictable RAM usage
✅ Better for huge models (>2GB checkpoints)
✅ Faster than f.write() (2-4x)
❌ Requires disk space (~2x checkpoint size)
❌ Additional I/O overhead (local + remote)

## Hybrid Optimization (Additional Approach)

```python
def atomic_save_checkpoints(checkpoint, path: str, threshold_mb: int = 2048):
    """
    Adaptive checkpoint saving based on checkpoint size.

    - < 2GB: Buffer-first with pipe() (fast, low overhead)
    - ≥ 2GB: Temp-file with fs.put() (stable, lower RAM)
    """
    import io
    import tempfile
    import os

    # Serialize to estimate size first
    buffer = io.BytesIO()
    torch.save(checkpoint, buffer)
    buffer.seek(0)
    size_mb = len(buffer.getvalue()) / (1024 * 1024)

    fs, urlpath = fsspec.core.url_to_fs(path)

    if size_mb < threshold_mb:
        # Fast path: pipe() for smaller checkpoints
        fs.pipe(urlpath, buffer.getvalue())
    else:
        # Stable path: temp file for large models
        with tempfile.NamedTemporaryFile(delete=False) as tmp:
            tmp.write(buffer.getvalue())
            tmp.close()
            fs.put(tmp.name, urlpath)
            os.unlink(tmp.name)
```

## Additional Optimizations

1. **Compression Tiering:**
```python
# For production models, enable PyTorch compression
torch.save(checkpoint, buffer, _use_new_zipfile_serialization=True)
```

2. **Async Upload with Verification:**
```python
from concurrent.futures import ThreadPoolExecutor

def save_async(checkpoint, path):
    with ThreadPoolExecutor() as executor:
        future = executor.submit(_atomic_save, checkpoint, path)
        # Verify upload in background
        future.add_done_callback(lambda f: verify_upload(path))
```

## Production Testing Recommendation

Test with:
- Small model: 100MB checkpoint (Approach 1 wins)
- Medium model: 1GB checkpoint (Approach 1 wins)
- Large model: 5GB checkpoint (Approach 2 wins)
- Huge model: 20GB+ checkpoint (Approach 2 necessary)

Would love to help benchmark and implement this! The hybrid approach gives the best of both worlds.

**Comment by c-pozzi** [comment]
Hi @deependujha, @ankitaluthra1 -- I ran extended benchmarks on AWS S3 to confirm this issue is not GCS-specific.

## Root cause

`_atomic_save` currently uses `fs.open(urlpath, "wb") + f.write()` which maps to a **single PUT request** on S3/GCS/Azure. This is capped at one TCP connection (~39 MB/s on EC2). `fs.pipe()` and `fs.put()` both trigger s3fs **multipart upload** under the hood, splitting the payload into parallel chunks and saturating available bandwidth.

Importantly, the `bytesbuffer` is already fully in memory before the upload starts (see `torch.save(checkpoint, bytesbuffer)` in `cloud_io.py`), so switching to `fs.pipe()` requires no architectural change -- the bytes are already there.

## S3 Benchmark Results

**Setup:** EC2 instance (us-east-1), Python 3.10, fsspec 2026.2.0, s3fs 2026.2.0, PyTorch 2.10.0. Checkpoints are `torch.randn` tensors serialized with `torch.save`. First run per method discarded as warmup to exclude TCP connection setup overhead. Remaining runs averaged.

| Size | `f.write()` (current) | `fs.pipe()` | `fs.put()` |
|------|----------------------|-------------|------------|
| 100 MB | 3.69s (27.1 MB/s) | 3.84s (26.1 MB/s) -- ~1x | 3.41s (29.3 MB/s) -- ~1x |
| 500 MB | 14.21s (35.2 MB/s) | 3.01s (166.2 MB/s) -- **4.7x** | 3.41s (146.6 MB/s) -- **4.2x** |
| 2 GB | 52.66s (38.0 MB/s) | 11.34s (176.4 MB/s) -- **4.6x** | 10.97s (182.3 MB/s) -- **4.8x** |
| 5 GB | 127.25s (39.3 MB/s) | 40.54s (123.4 MB/s) -- **3.1x** | 28.38s (176.2 MB/s) -- **4.5x** |
| 20 GB | 506.63s (39.5 MB/s) | 106.69s (187.4 MB/s) -- **4.7x** | 103.23s (193.8 MB/s) -- **4.8x** |

### Key findings

- `f.write()` plateaus at **~39 MB/s** from 2 GB onwards regardless of checkpoint size -- confirmed hard single-connection ceiling
- `fs.put()` delivers consistent **4.2-4.8x speedup** across all sizes above 500 MB with low variance
- `fs.pipe()` is competitive but shows higher variance at 5 GB (28-47s spread), making it less reliable for production training
- At 100 MB neither alternative offers a meaningful speedup -- multipart overhead cancels the parallelism benefit at small sizes
- Impact on a 20 GB checkpoint (realistic for large LLMs): **~507s today vs ~103s with `fs.put()`**, saving ~7 minutes per checkpoint save

<details>
<summary>Benchmark script</summary>

https://gist.github.com/9c06c39ce070a2cdf8f99bfcf7047484

</details>

## Proposed fix

Both `fs.pipe()` and `fs.put()` fix the bottleneck, but they have different tradeoffs:

**`fs.pipe()` -- recommended as default**

```python
# Before
fs, urlpath = fsspec.core.url_to_fs(str(filepath))
with fs.transaction, fs.open(urlpath, "wb") as f:
    f.write(bytesbuffer.getvalue())

# After
fs, urlpath = fsspec.core.url_to_fs(str(filepath))
with fs.transaction:
    fs.pipe(urlpath, bytesbuffer.getvalue())
```

- One-line change, no new dependencies
- Data is already in memory as `bytesbuffer` -- no disk I/O added
- 4.3-4.7x speedup at sizes >= 500 MB
- Works transparently across S3, GCS, and Azure via fsspec abstraction
- No behavior change for local filesystems

**`fs.put()` -- opt-in for maximum throughput**

```python
import tempfile

with tempfile.NamedTemporaryFile(delete=False) as tmp:
    tmp.write(bytesbuffer.getvalue())
    tmp_path = tmp.name
try:
    fs.put(tmp_path, urlpath)
finally:
    os.unlink(tmp_path)
```

- Consistently fastest option: 4.2-4.8x with lower variance than `fs.pipe()`
- Tradeoff: requires writing the full checkpoint to local disk first, meaning disk space >= checkpoint size must be available
- At 20 GB this is a real constraint -- not all training machines have sufficient free disk

**Recommendation:** use `fs.pipe()` as the default replacement. For users who want maximum throughput and can guarantee disk availability, `fs.put()` could be exposed as an option (e.g. via a `storage_options` flag in `TorchCheckpointIO`).

**Comment by c-pozzi** [comment]
This is a one-line change that works transparently across S3, GCS, and Azure via fsspec's abstraction. No behavior change for local filesystems (fsspec falls back to a standard write for `file://` paths).

> **Note:** Azure Blob Storage benchmark pending -- will update this when results available.

**Comment by c-pozzi** [comment]
**Update: Azure Blob Storage benchmark results + revised recommendation**

Following up on my S3 results above. I ran the same benchmark on Azure Blob Storage (Azure VM → Azure Blob, West Europe, Standard_D16s_v3, adlfs, fsspec 2026.2.0).

**Azure Blob Storage results:**

| Size | f.write() (current) | fs.pipe() | fs.put() |
|------|---------------------|-----------|----------|
| 100 MB | 0.76s (131 MB/s) | 2.23s (44.8 MB/s) — 0.34x | 2.09s (47.8 MB/s) — 0.36x |
| 500 MB | 3.06s (163 MB/s) | 11.04s (45.3 MB/s) — 0.28x | 10.94s (45.7 MB/s) — 0.28x |
| 2 GB | 11.42s (175 MB/s) | 42.85s (46.7 MB/s) — 0.27x | 43.10s (46.4 MB/s) — 0.27x |

**Azure behaves opposite to S3/GCS.** `fs.pipe()` and `fs.put()` are 3-4x slower than `f.write()` on Azure at every size tested. The pattern is consistent and does not reverse with size. Root cause is likely`adlfs` stages blocks sequentially rather than using true parallel multipart upload, so the single PUT path via `f.write()` is faster on Azure.

**Revised recommendation:**

The fix should be S3/GCS-specific. The existing `_is_object_storage()` helper currently groups S3, GCS, and Azure together — but our benchmarks show Azure needs different treatment. A minimal change would be adding an `exclude_azure` flag to the existing helper (default `False`, so all current callers are unaffected):
```python
# _is_object_storage with exclude_azure=True returns False for AzureBlobFileSystem
with fs.transaction:
    if _is_object_storage(fs, exclude_azure=True):
        fs.pipe(urlpath, bytesbuffer.getvalue())
    else:
        with fs.open(urlpath, "wb") as f:
            f.write(bytesbuffer.getvalue())
```

This gives 4.3-4.8x speedup on S3 and GCS with no regression on Azure or local filesystems. Happy to open a PR with this if the direction looks good.
@deependujha @lantiga @justusschock

**Comment by ankitaluthra1** [comment]
This is great ! Thankyou @c-pozzi for all your work on this and @deependujha for accepting this optimisation, When is the next release planned?  I'm looking forward to  try this out on my workloads once the new version drops, to see the real speedup !

**Comment by deependujha** [comment]
Hi @ankitaluthra1, we’ll soon be releasing 2.6.2, will try to include this PR in the release too.

**Comment by ankitaluthra1** [comment]
@deependujha Any ETA on the release containing this fix ?

**Comment by ankitaluthra1** [comment]
@justusschock This improvement seems to be excluded from recent release, any reason its not included ?
